### PR TITLE
[KED-1683] 08_advanced_io is missing `_exists()` method

### PR DESCRIPTION
**Pull request by WaylonWalker**
Oddly enough as I was making this edit I noticed that it was in the docstring of the AbstractVersionedDataSet.  This is a fairly easy change as it just replicates existing docs from https://github.com/quantumblacklabs/kedro/blob/16604ec0203432b9897abb34e3bb4d1c59d1f81d/kedro/io/core.py#L491

## Description

As I was looking for an example custom DataSet from the docs I noticed that there was not a `_exists` method on the dataset.  This definitely needs to be in the place where people are looking for example for.

## Development notes

It actually came from the `AbstractVersionedDataSet` docstring.

## Checklist

- [x] Read the [contributing](../CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](../RELEASE.md) file
- [NA] Added tests to cover my changes

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
- This is my own ideas, not tied to my employer

### CODE REVIEW

**Comment by WaylonWalker** [comment]
I just took a second look at this and noticed that `param1` and `param2` are not optional.  Should they be optional or included in the example yaml.  personally I think they should be in the yaml for a more complete example.


``` yaml
my_dataset:
  type: <path-to-my-own-dataset>.MyOwnDataSet
  filepath: data/01_raw/my_data.csv
  versioned: true
  param1: <param1-value>
  param2: <param2-value>
```

**Comment by deepyaman** [comment]
Overriding `_exists` is optional (as mentioned just above in the Markdown, will fall back to `_local_exists` I think), but I do kinda appreciate the consistency between the docstring and the docs...

File: RELEASE.md [hunk 1]
```diff
@@ -47,6 +47,7 @@
 * Added option to lint the project without applying the formatting changes (`kedro lint --check-only`).
 * Improved the error message for `DataSetNotFoundError` to suggest possible dataset names user meant to type.
 * Replaced `functools.lru_cache` with `cachetools.cachedmethod` in `PartitionedDataSet` and `IncrementalDataSet` for per-instance cache invalidation.
+* Added missing `_exists` method to `MyOwnDataSet` example in 04_user_guide/08_advanced_io.
```
Comment (lorenabalan) [inline RELEASE.md:50]: We've reshufled the contents of this file in preparation for the upcoming release. Feel free to move this under "docs improvements" bullet points, or remove altogether.

File: docs/source/04_user_guide/08_advanced_io.md [hunk 2]
```diff
@@ -84,6 +88,8 @@ my_dataset:
   type: <path-to-my-own-dataset>.MyOwnDataSet
   filepath: data/01_raw/my_data.csv
   versioned: true
+  param1: <param1-value>
+  param2: <param2-value>
```
Comment (lorenabalan) [inline docs/source/04_user_guide/08_advanced_io.md:92]: Maybe make one parameter optional and one mandatory, for completeness?

File: docs/source/04_user_guide/08_advanced_io.md [hunk 3]
```diff
@@ -73,6 +73,10 @@ class MyOwnDataSet(AbstractVersionedDataSet):
         save_path = self._get_save_path()
         df.to_csv(save_path)
 
+    def _exists(self) -> bool:
+        path = self._get_load_path()
+        return path.is_file()
```
Comment (lorenabalan) [inline docs/source/04_user_guide/08_advanced_io.md:78]: Well spotted! I just noticed the `super()` call is also out of date, it should be passing a purepath, e.g. `PurePosixPath(filepath)`. Do you mind making that change here on line 64, and in `kedro/io/core.py` - line 479 or thereabout? 

Personal preference, I would also leave filepath and version as the first args, then followed by `param1` and an optional `param2`.

Comment (WaylonWalker) [reply docs/source/04_user_guide/08_advanced_io.md:78]: 💯 Agree, I like to keep arguments in the natural order they get called.  Gives me goosebumps that someone is going to not use keyword arguments and call it backwards.

File: docs/source/04_user_guide/08_advanced_io.md (hunk 2 above)
Comment (WaylonWalker) [reply docs/source/04_user_guide/08_advanced_io.md:92]: param2 is now `True` by default

File: kedro/io/core.py [hunk 4]
```diff
@@ -106,19 +106,30 @@ class AbstractDataSet(abc.ABC):
...
-        >>>         return dict(param1=self._param1, param2=self._param2)
+        >>>         return dict(    Example catalog.yml specification
+
+    Example catalog.yml specification
+    ::
+        my_dataset:
+            type: <path-to-my-own-dataset>.MyOwnDataSet
+            filepath: data/01_raw/my_data.csv
+            param1: <param1-value> # param1 is a required argument
+            # param2 will be True by defaultparam1=self._param1, param2=self._param2)
```
Comment (lorenabalan) [inline kedro/io/core.py:132]:
```suggestion
            # param2 will be True by default
```

File: kedro/io/core.py [hunk 5]
```diff
@@ -106,19 +106,30 @@ class AbstractDataSet(abc.ABC):
...
         >>>     def _save(self, df: pd.DataFrame) -> None:
-        >>>         print("Dummy save: {}".format(self._param2))
+        >>>         df.to_csv(str(self._filepath))
+        >>>
+        >>>     def _exists(self) -> bool:
+        >>>         return path._filepath.is_file()
         >>>
         >>>     def _describe(self):
-        >>>         return dict(param1=self._param1, param2=self._param2)
+        >>>         return dict(    Example catalog.yml specification
```
Comment (lorenabalan) [inline kedro/io/core.py:124]:
```suggestion
        >>>         return dict(param1=self._param1, param2=self._param2)
```

File: RELEASE.md [hunk 6]
```diff
@@ -53,6 +54,8 @@
 * Improved the error message for `DataSetNotFoundError` to suggest possible dataset names user meant to type.
 * Added the option for contributors to run Kedro tests locally without Spark installation with `make test-no-spark`.
 * Added option to lint the project without applying the formatting changes (`kedro lint --check-only`).
+* Improved the error message for `DataSetNotFoundError` to suggest possible dataset names user meant to type.
+* Replaced `functools.lru_cache` with `cachetools.cachedmethod` in `PartitionedDataSet` and `IncrementalDataSet` for per-instance cache invalidation.
```
Comment (lorenabalan) [inline RELEASE.md:58]: Looks like a merge conflict residue, the lines have been moved above. :)
```suggestion
```

File: docs/source/04_user_guide/08_advanced_io.md [hunk 7]
```diff
@@ -52,16 +52,16 @@ In order to enable versioning, you need to update the `catalog.yml` config file
 An example dataset could look similar to the below:
 
 ```python
-from pathlib import Path
+from pathlib import PurePath
```
Comment (lorenabalan) [inline docs/source/04_user_guide/08_advanced_io.md:55]: We normally use `PurePosixPath` in concrete datasets 🤔 but maybe `PurePath` is also ok
```suggestion
from pathlib import PurePosixPath
```

File: kedro/io/core.py (hunk 4 above)
Comment (WaylonWalker) [reply kedro/io/core.py]: Good Catch, not sure what happened there.

**Comment by WaylonWalker** [comment]
@lorenabalan I copied the examples to a local editor and noticed that `is_file()` was not  a method on `PurePosixPath` so I converted the check to `Path(self._filepath).exists()` and `Path(path).exists()`.  Is this the correct method?  I don't often use the `PurePosixPath`.

**Comment by lorenabalan** [comment]
> @lorenabalan I copied the examples to a local editor and noticed that `is_file()` was not a method on `PurePosixPath` so I converted the check to `Path(self._filepath).exists()` and `Path(path).exists()`. Is this the correct method? I don't often use the `PurePosixPath`.

That's perfectly fine for now - we rely on the `fsspec` to tell us if a path string exists or not, but that's too much detail for a docstring.

File: kedro/io/core.py [hunk 8]
```diff
@@ -101,24 +101,36 @@ class AbstractDataSet(abc.ABC):
...
+        >>>
+        >>>     def _exists(self) -> bool:
+        >>>         return Path(self._filepath).exists()
         >>>
         >>>     def _describe(self):
         >>>         return dict(param1=self._param1, param2=self._param2)
+
+    Example catalog.yml specification
+    ::
+        my_dataset:
```
Comment (mzjp2) [inline kedro/io/core.py:129]: Does this actually render in the docs? I thought you'd need another blank line.

File: kedro/io/core.py [hunk 9]
```diff
@@ -490,10 +503,19 @@ class AbstractVersionedDataSet(AbstractDataSet, abc.ABC):
...
         >>>         path = self._get_load_path()
-        >>>         return path.is_file()
+        >>>         return Path(path).exists()
         >>>
         >>>     def _describe(self):
         >>>         return dict(version=self._version, param1=self._param1, param2=self._param2)
+
+    Example catalog.yml specification
+    ::
+        my_dataset:
```
Comment (mzjp2) [inline kedro/io/core.py:513]: similar re: rendering in docs and blank line between `::` and `my_dataset`.

File: kedro/io/core.py (hunk 8 above)
Comment (WaylonWalker) [reply kedro/io/core.py:129]: Thanks @mzjp2!  It's been a long time since I have used restructured text.

Comment (WaylonWalker) [reply kedro/io/core.py:131]: I think I got it fixed.  popped it into an online renderer and it seemed to like the latest change.

File: kedro/io/core.py (hunk 9 above)
Comment (WaylonWalker) [reply kedro/io/core.py]: I think I got it fixed.  popped it into an online renderer and it seemed to like the latest change.

**Comment by lorenabalan** [comment]
@WaylonWalker I'll quickly take over this to get it merged hope that's okay

**Pull request merged**